menhir-recover/recovery_new.ml

```diff
@@ -0,0 +1,196 @@
+open MenhirSdk.Cmly_api
+open Synthesis
+ 
+module type RECOVERY = sig
+  module G : GRAMMAR
+
+  type item = G.lr1 * G.production * int
+
+  type recovery = {
+    prefix: int;
+    cases: (G.lr1 option * item list) list;
+  }
+  (** [prefix] is the size of the known prefix of the stack.
+      It means that in the kernel of current state, there is an item whose dot
+      is at position [prefix].
+      (we know the incoming symbols for these stack frames and we can enumerate
+      the possible state numbers).
+
+      [cases] is a mapping that associates to each possible state found at
+      stack.[-prefix]
+      (or None if the stack is empty) a list of reductions to execute.
+
+      The actual list of actions to reduce an item [(state, prod, pos)] is
+      given by
+          [Synthesizer.solution (Trail (state, prod, pos))]
+    *)
+
+  val recover : G.lr1 -> recovery
+  val report : Format.formatter -> unit
+end
+ 
+module Recover (G : GRAMMAR) (S : SYNTHESIZER with module G := G)
+  : RECOVERY with module G := G =
+struct
+  open G
+  open Utils
+
+  type item = lr1 * production * int
+
+  type recovery = {
+    prefix: int;
+    cases: (G.lr1 option * item list) list;
+  }
+
+  type trace = { cost : Cost.t; items : item list }
+
+  module Trace = struct
+    type t = trace
+    let min tr1 tr2 =
+      Cost.arg_min (fun t -> t.cost) tr1 tr2
+
+    let cat tr1 tr2 =
+      { cost = Cost.add tr1.cost tr2.cost; items = tr1.items @ tr2.items }
+  end
+
+  let depth = Lr1.tabulate (fun st ->
+    List.fold_left (fun pos (_, pos') -> max pos pos') 0 @@ Lr0.items (Lr1.lr0 st))
+
+  module State = struct
+    module S = Set.Make (
```

Comment:
`S` is a bad abbreviation because it often stands for `S`ignature.
I would rather use `Set` or `NS` which stands for Nonterminal Set or `NSet`/`NTSet`